internal/gkr/gkrtypes/types.go

```diff
@@ -0,0 +1,425 @@
+		NbInstances:  info.NbInstances,
+		HashName:     info.HashName,
+		Dependencies: info.Dependencies,
+		Prints:       info.Prints,
+	}, err
+}
+
+// WireAssignment is assignment of values to the same wire across many instances of the circuit
+type WireAssignment []polynomial.MultiLin
+
+func (a WireAssignment) Permute(p gkrinfo.Permutations) {
+	utils.Permute(a, p.WiresPermutation)
+	for i := range a {
+		if a[i] != nil {
+			utils.Permute(a[i], p.InstancesPermutation)
+		}
+	}
+}
+
+func (a WireAssignment) NbInstances() int {
+	for _, aW := range a {
+		if aW != nil {
+			return len(aW)
+		}
+	}
+	panic("empty assignment")
+}
+
+func (a WireAssignment) NbVars() int {
+	for _, aW := range a {
+		if aW != nil {
+			return aW.NumVars()
+		}
+	}
+	panic("empty assignment")
+}
+
+// ProofSize computes how large the proof for a circuit would be. It needs NbUniqueOutputs to be set.
+func (c Circuit) ProofSize(logNbInstances int) int {
+	nbUniqueInputs := 0
+	nbPartialEvalPolys := 0
+	for i := range c {
+		nbUniqueInputs += c[i].NbUniqueOutputs // each unique output is manifest in a finalEvalProof entry
+		if !c[i].NoProof() {
+			nbPartialEvalPolys += c[i].Gate.Degree() + 1
+		}
+	}
+	return nbUniqueInputs + nbPartialEvalPolys*logNbInstances
+}
+
+// makeNeg1Slice returns a slice of size n with all elements set to -1.
+func makeNeg1Slice(n int) []int {
+	res := make([]int, n)
+	for i := range res {
+		res[i] = -1
+	}
+	return res
+}
+
+type topSortData struct {
+	outputs    [][]int
+	status     []int // status > 0 indicates number of inputs left to be ready. status = 0 means ready. status = -1 means done
+	leastReady int
+}
+
+func (d *topSortData) markDone(i int) {
+
+	d.status[i] = -1
+
+	for _, outI := range d.outputs[i] {
+		d.status[outI]--
+		if d.status[outI] == 0 && outI < d.leastReady {
+			d.leastReady = outI
+		}
+	}
+
+	for d.leastReady < len(d.status) && d.status[d.leastReady] != 0 {
+		d.leastReady++
+	}
+}
+
+func statusList(c Circuit) []int {
+	res := make([]int, len(c))
+	for i := range c {
+		res[i] = len(c[i].Inputs)
+	}
+	return res
+}
+
+// TopologicalSort sorts the wires in order of dependence. Such that for any wire, any one it depends on
+// occurs before it. It tries to stick to the input order as much as possible. An already sorted list will remain unchanged.
+// It also sets the nbOutput flags, and a dummy IdentityGate for input wires.
+// Worst-case inefficient O(n^2), but that probably won't matter since the circuits are small.
+// Furthermore, it is efficient with already-close-to-sorted lists, which are the expected input
+func (c Circuit) TopologicalSort() []*Wire {
+	var data topSortData
+	data.outputs = c.OutputsList()
+	data.status = statusList(c)
+	sorted := make([]*Wire, len(c))
+
+	for data.leastReady = 0; data.status[data.leastReady] != 0; data.leastReady++ {
+	}
+
+	for i := range c {
+		sorted[i] = &c[data.leastReady]
+		data.markDone(data.leastReady)
+	}
+
+	return sorted
+}
+
+type errorString string
```

Comment:
Also, why this? You can just do
```
var ErrZeroFunction = errors.New("detected a zero function")
```
It is very ideomatic Go and whole standard library uses that approach instead: https://pkg.go.dev/io/fs#pkg-variables